src/runtime/io.cpp

```diff
@@ -630,6 +631,165 @@ extern "C" LEAN_EXPORT obj_res lean_io_prim_handle_put_str(b_obj_arg h, b_obj_ar
+    }
+
+    UCalendar *cal = ucal_open(tzID, -1, NULL, UCAL_GREGORIAN, &status);
+
+    if (U_FAILURE(status)) {
+        ucal_close(cal);
+        return lean_io_result_mk_error(lean_decode_io_error(EINVAL, mk_string("failed to open calendar")));
+    }
+
+    int64_t timestamp_secs = lean_scalar_to_int64(tm_obj);
+
+    ucal_setMillis(cal, timestamp_secs * 1000, &status);
+    if (U_FAILURE(status)) {
+        ucal_close(cal);
+        return lean_io_result_mk_error(lean_decode_io_error(EINVAL, mk_string("failed to set calendar time")));
+    }
+    
+    int32_t dst_offset = ucal_get(cal, UCAL_DST_OFFSET, &status);
+
+    if (U_FAILURE(status)) {
+        ucal_close(cal);
+        return lean_io_result_mk_error(lean_decode_io_error(EINVAL, mk_string("failed to get dst_offset")));
+    }
+
+    int is_dst = dst_offset != 0;
+
+    UDate nextTransition;
+    if (!ucal_getTimeZoneTransitionDate(cal, UCAL_TZ_TRANSITION_NEXT, &nextTransition, &status)) {
+        ucal_close(cal);
+        return io_result_mk_ok(mk_option_none());
+    }
+
+    if (U_FAILURE(status)) {
+        ucal_close(cal);
+        return lean_io_result_mk_error(lean_decode_io_error(EINVAL, mk_string("failed to get next transation")));
+    }
+
+    int32_t tzIDLength = ucal_getTimeZoneDisplayName(cal, is_dst ? UCAL_DST : UCAL_STANDARD, "en_US", tzID, 32, &status);
+    
+    if (U_FAILURE(status)) {
+        ucal_close(cal);
+        return lean_io_result_mk_error(lean_decode_io_error(EINVAL, mk_string("failed to timezone identifier")));
+    }
+
+    char dst_name[256];
+    u_strToUTF8(dst_name, sizeof(dst_name), NULL, tzID, tzIDLength, &status);
+
+    UChar display_name[32];
+    int32_t display_name_len = ucal_getTimeZoneDisplayName(cal, is_dst ? UCAL_SHORT_DST : UCAL_SHORT_STANDARD, "en_US", display_name, 32, &status);
+
+    if (U_FAILURE(status)) {
+        ucal_close(cal);
+        return lean_io_result_mk_error(lean_decode_io_error(EINVAL, mk_string("failed to read abbreaviation")));
+    }
+
+    char display_name_str[256];
+    u_strToUTF8(display_name_str, sizeof(display_name_str), NULL, display_name, display_name_len, &status);
+
+    if (U_FAILURE(status)) {
+        ucal_close(cal);
+        return lean_io_result_mk_error(lean_decode_io_error(EINVAL, mk_string("failed to get abbreviation to cstr")));
+    }
+
+    int32_t zone_offset = ucal_get(cal, UCAL_ZONE_OFFSET, &status);
+    zone_offset += dst_offset;
+
+    if (U_FAILURE(status)) {
+        ucal_close(cal);
+        return lean_io_result_mk_error(lean_decode_io_error(EINVAL, mk_string("failed to get zone_offset")));
+    }
+
+    int64_t tm = (int64_t)(nextTransition / 1000.0);
+    ucal_close(cal);
+
+    int offset_seconds = zone_offset / 1000;
+
+    lean_object *lean_tz = lean_alloc_ctor(0, 3, 1);
+    lean_ctor_set(lean_tz, 0, lean_int_to_int(offset_seconds));
+    lean_ctor_set(lean_tz, 1, lean_mk_ascii_string_unchecked(dst_name));
```

Comment:
Looks like you're hoping that a time zone name will never be non-ascii. It might be better to play it safe and use `lean_mk_string_from_bytes_unchecked` here.